# Data wrangling on [people_data.csv](https://piazza.com/redirect/s3?bucket=uploads&prefix=attach%2Fjsagy1god6f2bq%2Fjdl2dukjx9e4n9%2Fjswf28nz9jht%2Fpeople_data.csv) by `Mostafa Tarek` using `MongoDB`, `Pymongo` 

__vis the data and assessing it__

In [2]:
import pandas as pd 
df = pd.read_csv('./people_data.csv')
df

ID     firstname      lastname   age  \
0        0         magdi  abd elrahman  35.0   
1        1         samir         tamer  84.0   
2        2         eslam        rashad  74.0   
3        3           ali         eslam   NaN   
4        4          ehab         maher   NaN   
5        5         yosra       ibrahim   NaN   
6        6          mona         Arshd   NaN   
7        7         yosra       ibrahim  38.0   
8        8           mai         mazen   NaN   
9        9         sanaa          samy   NaN   
10      10        rashad          amer  33.0   
11      11         shady         Arshd  69.0   
12      12       ibrahim         tamer  21.0   
13      13  abd elrahman         Arshd  31.0   
14      14         tamer          ehab  59.0   
15      15         yosra         hosny   NaN   
16      16       ibrahim         hatem  58.0   
17      17       ibrahim     abd allah  50.0   
18      18         tamer         hatem  22.0   
19      19          eyad          rami   NaN   
20      20       mohamed          samy  36.0   
21      21         amera         tarek  30.0   
22      22          rana          omar   NaN   
23      23           amr     abd allah  73.0   
24      24          rana     abd allah  76.0   
25      25         sameh       mohamed   NaN   
26      26          amar     abd allah  38.0   
27      27          omar         hatem  42.0   
28      28         hatem       ibrahim  59.0   
29      29         Arshd          omar  88.0   
...    ...           ...           ...   ...   
9970  9970         sameh           amr  57.0   
9971  9971           ali          amar   NaN   
9972  9972          mona       mahmoud  61.0   
9973  9973         Arshd          rami  75.0   
9974  9974           mai           ali  31.0   
9975  9975         ahmed         eslam   NaN   
9976  9976         tamer         mazen  53.0   
9977  9977          rana         hosny  35.0   
9978  9978        samera           amr   NaN   
9979  9979         amera       mahmoud  31.0   
9980  9980       nourhan       mahmoud  35.0   
9981  9981       nourhan          ehab  88.0   
9982  9982           mai       mahmoud  80.0   
9983  9983         mazen          eyad  18.0   
9984  9984         salma          omar  21.0   
9985  9985         yosra           amr  23.0   
9986  9986         Arshd           amr  80.0   
9987  9987       mohamed          omar   NaN   
9988  9988        samera          amar  88.0   
9989  9989         tamer         sameh  41.0   
9990  9990         maher        rashad  76.0   
9991  9991          rana        rashad   NaN   
9992  9992          amar       ibrahim  33.0   
9993  9993           aya         samir  81.0   
9994  9994         salma          amer   NaN   
9995  9995         amera         hatem  58.0   
9996  9996          rana          alaa  89.0   
9997  9997           mai         mazen  63.0   
9998  9998         shady           ali   NaN   
9999  9999           aya         magdi  35.0   

                                             currentJob  \
0                     Surveying and Mapping Technicians   
1                                                   NaN   
2                                   Mapping Technicians   
3     Health and Safety Engineers, Except Mining Saf...   
4                                   Photonics Engineers   
5                                  Validation Engineers   
6                                Mechatronics Engineers   
7                                      Energy Engineers   
8                                                   NaN   
9                               Environmental Engineers   
10                Marine Engineers and Naval Architects   
11                                Biochemical Engineers   
12                                   Chemical Engineers   
13                                Wind Energy Engineers   
14                   Industrial Engineering Technicians   
15                       Electro-Mechanical Technicians   
16                          

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 10 columns):
ID             10000 non-null int64
firstname      10000 non-null object
lastname       10000 non-null object
age            7181 non-null float64
currentJob     8732 non-null object
prevJobs       6039 non-null object
friendsID      9734 non-null object
hobbies        6045 non-null object
email          5876 non-null object
nationality    10000 non-null object
dtypes: float64(1), int64(1), object(8)
memory usage: 781.3+ KB


In [4]:
from pymongo import MongoClient
client = MongoClient('mongodb://localhost:27017')

__create a `people` database and assign it to `peopleDb` var

In [5]:
peopleDb = client["people"]

In [6]:
client.list_database_names()

['admin', 'config', 'local', 'people']

__add `people_info` collection to `people` database__ and assign it to `peopleInfo`

In [7]:
peopleInfo = peopleDb["people_info"]

In [8]:
peopleDb.list_collection_names()

['people_info']

__assign the columns names to `columns` and remove the df after assessing and vis it to save memory__


In [9]:
columns = df.columns
del df 

In [10]:
import csv  
from pprint import pprint 

In [11]:
peopleInfo.delete_many({})
print("count before inserting {} \n inserting... ".format(peopleInfo.count_documents({})))
with open('./people_data.csv','r') as file :
    reader = csv.DictReader(file, fieldnames=columns)
    next(reader) # skipt the header 
    data = [] 
    for row in reader:
        data.append(dict(row)) 
peopleInfo.insert_many(data)
print("count after inserting {} ".format(peopleInfo.count_documents({})))

count before inserting 0 
 inserting... 
count after inserting 10000 


### 1- remove all fields with empty values 

In [12]:
c = 0 
for i in columns:
    c += peopleInfo.count_documents({i:{"$in":["NaN","nan"," ","",None,"Null","null"]}})
    peopleInfo.delete_many({i:{"$in":["NaN","nan"," ","",None,"Null","null"]}})
print("deleted {} documents".format(c))

deleted 8703 documents


In [13]:
#uncomment to vis the docs 
# for doc in peopleInfo.find({}):
#     pprint(doc)
#     pprint('\n')

### 2- change `prevjops` ,` hobbies` and `friendsID`  to array data type

In [14]:
for field in ["prevJobs" , "hobbies" , "friendsID"]:
    for doc in peopleInfo.find({field: {"$exists":1}}):
        newvalues = { "$set": {field: doc[field].split(',')}}
        peopleInfo.update_one({"_id":doc["_id"]}, newvalues)

In [15]:
#uncomment to vis 
# for doc in peopleInfo.find({}):
#     print(doc["prevJobs"])
#     print('\n')

### 3- Find all persons’ document that his `current job` is `his first job` and also his `age` between `20` and` 30 year`

__convet `age` to `float` so we can manipulate  with it__

In [16]:
type(next(peopleInfo.find( { "age": { "$exists": 1 } } ))["age"])

str

In [17]:
for doc in peopleInfo.find( { "age": { "$exists": 1 } } ):
    peopleInfo.update_one({"_id":doc["_id"]},{"$set":{"age":float(doc["age"])}})

In [18]:
type(next(peopleInfo.find( { "age": { "$exists": 1 } } ))["age"])

float

In [19]:
for doc in peopleInfo.find({ "$and":[{"currentJob":{"$exists":1}},{"prevJobs":{"$exists":1}},  { "age": { "$lte": 30 } }, { "age": {"$gte":20} }  ] }):
        if doc["currentJob"] in doc["prevJobs"]:
            #uncomment to vis 
            #print(doc)
            pass

## 4- for each email domain find how many persons using it?
__Email domains : yahoo , gmail , hotmail ….etc__

In [20]:
for i in peopleInfo.aggregate([
  { "$project" : { "emailDomains" : { "$split": ["$email","@"] }} },
    {"$project": { "emailDomains": {"$slice":["$emailDomains",-1]}}},
    {"$unwind": "$emailDomains"}, #so we can split it by . at the next pipline as the [] type will raise an error, in other words this step get the string from an array 
   { "$project" : { "emailDomains" : { "$split": ["$emailDomains","."] }} },
    {"$project": { "emailDomains": {"$slice":["$emailDomains",1]}, "qt":1}},
    {"$unwind": "$emailDomains"},
  { "$group" : { "_id": "$emailDomains" , "count" : { "$sum" : 1 } } },
  { "$sort" : { "total_qty" : -1 } }
]):
    print(i)

{'_id': 'hotmail', 'count': 446}
{'_id': 'yahoo', 'count': 413}
{'_id': 'gmail', 'count': 438}


## 5- for all persons with fullname : `mohamed samir` find all documents of their friends .

In [21]:
friendsDocs = [] 
for doc in peopleInfo.aggregate([
    {"$project": {"full_name": {"$concat": ["$firstname"," ","$lastname"]}, "friendsID":"$friendsID"}},
    {"$unwind":"$friendsID"},
    {"$match": {"full_name": "mohamed samir"}}
]):
    friendsDocs.append(peopleInfo.find({"ID":doc["friendsID"]}))
"number of docs is  {} docs ".format(len(friendsDocs))

'number of docs is  22 docs '

In [22]:
#uncomment to vis 
# for i in friendsDocs:
#     for j in i:
#         print(j)

- note that the above output has just `two` docs out of `22` docs as we preveously removed the docs that contain empty fields at procedure number `one` `1- remove all fields with empty values ` 

### 6.1- for all Egyptian persons , how many persons with first name `ahmed`  ? 

In [24]:
for i in peopleInfo.aggregate([
    { "$match": {"firstname":"ahmed", "nationality":"egyptian"}},
    { "$group": {"_id": "$firstname" , "count" : { "$sum" : 1 } } }
]):
    print(i["count"])

23


###  6.2- how many persons `have more than one hobby` ?

In [29]:
a =  peopleInfo.aggregate([
    {"$match": { "nationality":"egyptian"}},
    {"$project":{"hobbiesCount":{"$size":"$hobbies"}}},
    {"$match": {"hobbiesCount": {"$gt":1}} }
])
len(list(a))

116

__but there is a problem that there maybe are some redundant persons so we should group by name at the end__
### so:

In [30]:
a =  peopleInfo.aggregate([
    {"$match": { "nationality":"egyptian"}},
    {"$project":{"hobbiesCount":{"$size":"$hobbies"},  "name":{"$concat":["$firstname"," ","$lastname"]}  }},
    {"$match": {"hobbiesCount": {"$gt":1}} },
    {"$group": {"_id":"$name"}},
])
len(list(a))

109

__note there were 116 - 109 == `7` redundant persons name__

###  6.3- and  find the `top 5 jobs with respect to number of people working it` 

In [33]:
list(peopleInfo.aggregate([
    {"$match": { "nationality":"egyptian"}},
    {"$group": {"_id":"$currentJob", "count":{"$sum":1}}},
    {"$sort": {"count":-1}},
    {"$limit":5}
]))

[{'_id': 'Architects, Except Landscape and Naval', 'count': 8},
 {'_id': 'Engineers, All Other', 'count': 7},
 {'_id': 'Architectural Drafters', 'count': 7},
 {'_id': 'Human Factors Engineers and Ergonomists', 'count': 7},
 {'_id': 'Non-Destructive Testing Specialists', 'count': 7}]

# conclusion 

### Personally, I do not see any advantage for MongoDB to make it overcome SQL "in this particular task" 